server/script/core/code-action.lua

```diff
@@ -137,7 +142,222 @@ local function changeVersion(uri, version, results)
+            if child.type == 'type.library' and child.value.uri == uri then
+                table.insert(path, child)
+                return path
+            end
+
+            if child.type == 'type.library' and child.value.child then
+                table.insert(path, child)
+                local result = findPath(uri, child, path)
+                if result then
+                    return result
+                end
+                table.remove(path)
+            end
+        end
+    end
+end
+
+local function getSafeIndexer(name)
+    if name:match('^%D[%w_]*$') then
+        return '.' .. name
+    else
+        return string.format('[%q]', name)
+    end
+end
+
+local function canIndexObject(object)
+    if not config.config.diagnostics.importScriptChildren then
+        if object.value[1]:match("Script$") then
+            return false
+        end
+    end
+
+    return true
+end
+
+local function findRelativeLuaPath(sourcePath, targetPath)
+    local sourcePathMap = {}
+    for index, node in ipairs(sourcePath) do
+        -- We only want to use relative paths if they don't go through game or a
+        -- service
+        local isGame = node == rbxlibs.global.game
+        local isChildOfGame = node.value.child and node.value.child.Parent == rbxlibs.global.game
+        if not isGame and not isChildOfGame then
+            sourcePathMap[node] = index
+        end
+    end
+
+    local commonAncestor
+    local commonAncestorTargetIndex
+    for index = #targetPath, 1, -1 do
+        local node = targetPath[index]
+        if sourcePathMap[node] then
+            commonAncestor = node
+            commonAncestorTargetIndex = index
+            break
+        end
+    end
+
+    if not commonAncestor then
+        return
+    end
+
+    local builder = {'script'}
+
+    for _ = #sourcePath - 1, sourcePathMap[commonAncestor], -1 do
+        table.insert(builder, '.Parent')
+    end
+
+    for index = commonAncestorTargetIndex + 1, #targetPath do
+        local node = targetPath[index]
+        
+        if index ~= #targetPath and not canIndexObject(node) then
+            return
+        end
+
+        table.insert(builder, getSafeIndexer(node.name))
+    end
+
+    return table.concat(builder)
+end
+
+local function getAbsoluteLuaPath(path)
+    local builder = { path[1].name }
+
+    for index, node in ipairs(path) do
+        if index ~= 1 then
+            if not canIndexObject(node) then
+                return
+            end
+
+            table.insert(builder, getSafeIndexer(node.name))
+        end
+    end
+
+    return table.concat(builder)
+end
+
+local function solveUndefinedGlobalImport(uri, diag, results)
+    local ast    = files.getAst(uri)
+    local offset = files.offsetOfWord(uri, diag.range.start)
+    local name = guide.eachSourceContain(ast.ast, offset, function (source)
+        if source.type ~= 'getglobal' then
+            return
+        end
+
+        return guide.getKeyName(source)
+    end)
```

Comment:
This is duplicate code from `solveUndefinedGlobal`

Reply:
This is now its own diagnostic, so I've left this duplicated. There's similar duplication between `solveUndefinedGlobal` and `solveLowercaseGlobal`. It's a small enough bit of code that I don't think it benefits much from being de-duplicated.

If we do de-duplicate it though, we should do it across this whole file -- maybe a function like `getKeyNameAtOffset(ast, offset, sourceType)`?

Reply:
That's ok then.